ARCs/arc-0052.md

```diff
@@ -0,0 +1,210 @@
+    letter: {
+      type: "string",
+    },
+  },
+};
+
+const metadata: SignMetadata = {
+  encoding: Encoding.BASE64,
+  schema: jsonSchema,
+};
+
+const signature: Uint8Array = await cryptoService.signData(
+  KeyContext.Address,
+  0,
+  0,
+  encoded,
+  metadata
+);
+expect(signature).toHaveLength(64);
+```
+
+##### ECDH (Elliptic Curve Diffie-Hellman)
+
+This operation is used to perform a secret key aggrement between two parties. This is useful for example for deriving a shared symmetric secret that is only known between the participants.
+
+The symmetric secret can be used for encrypting and decrypting messages.
+
+- **method name**: `ECDH`
+- **parameters**:
+  - `context`: The context of the key to be used for signing
+  - `account`: The account index to be used for key derivation. The value value should be hardened as per BIP44
+  - `keyIndex`: The key index to be used for key derivation.
+  - `otherPartyPub`: The public key of the other party
+- **returns**: Byte array containing the shared secret. The shared secret is a 32-byte array.
+
+###### Example
+
+```ts
+const aliceKey: Uint8Array = await cryptoService.keyGen(
+  KeyContext.Address,
+  0,
+  0
+);
+
+const bobKey: Uint8Array = await cryptoService.keyGen(KeyContext.Address, 0, 1);
+
+const aliceSharedSecret: Uint8Array = await cryptoService.ECDH(
+  KeyContext.Address,
+  0,
+  0,
+  bobKey
+);
+
+const bobSharedSecret: Uint8Array = await cryptoService.ECDH(
+  KeyContext.Address,
+  0,
+  1,
+  aliceKey
+);
+
+// Same secret can be used to encrypt by Alice and decrypt by Bob
+//and vice-versa
+expect(aliceSharedSecret).toEqual(bobSharedSecret);
+```
+
+## Rationale
+
+At it's core, blockchain keys and signatures are just implementations of cryptographic primitives. These primitives can be used for a wide variety of use cases, not just for the purpose of signing transactions. And wallets being applications that contain or serve as Key Management Systems (KMS), should be able to support a multiple of other use cases that are not just related to transactions. This creates the possibility of attaching on-chain behavior with off-chain behavior, allowing for example to concepts of Identity and Authentication.
+
+The current wallet APIs available only support key derivation for addresses and signing for transactions. In the broader context of Algorand and web3 ecosystems, there is a need for a more flexible API that allows for contextual key generation and signing. An example of this could be identity dedicated keys (i.e DIDs), or passkeys for authentication.
+
+## Reference Implementation
+
+Reference Implementation is included in the `assets` folder <a href= "https://github.com/algorandfoundation/ARCs/tree/main/assets/arc-0052">here</a>.
+
+## Security Considerations
+
+As this functionality is based on EdDSA, BIP32 HD derivations and ECDH, the security considerations of those specifications apply. Particularly, the Ed25519 derived keys must be clamped to avoid known attacks 
```

Comment:
Since we are including an ECDH function to produce a shared secret we should point out that the shared secret should NOT be used directly for encryption. Instead the output of ECDH should be concatenated with the bytes of the public keys and then hashed to produce a new secret.

> secret = Hash(concatenation(ECDH(skA, pkB), pkA, pkB))

(Where skA and pkA are of course produced following the BIP39/ARC-52 standard.)

Note that Alice and Bob need to agree beforehand on whose public key is concatenated first and whose second.

An explanation for why is given here: https://github.com/jedisct1/libsodium/issues/586#issuecomment-325182090

Scalarmult produces a point, a group element. While we use 32 bytes, corresponding to 2^256 possible values, in truth the order of the group is roughly ~2^252 and the output is not "uniformly random".

Hashing (e.g. with SHA512/256) makes the output uniformly random, but while the output would occupy the full 32 bytes space the range of inputs to the hash function is still constrained by ~2^252 possible inputs and the same number of outputs since hashes map (input) data to (output) value. By concatenating with something extra  increase the range of outputs.

Keep in mind that the different combinations of n,p in scalarmult produce the same new point as well. Since Alice's pk is x*G and Bob's y*G, their shared secret is xy*G. x=10, y = 50 would produce the same shared secret as x=20 and y = 25. However, Hash(Concat(150*G, 10*G, 50*G)) !== Hash(Concat(150*G, 20*G, 25*G)). So the space of potential secrets is boosted.

**Alternatively** we should update the ECDH so it doesn't return scalarmult(skA, pkB) directly. Instead we update it with an additional parameter "input order" and then returns either Hash(Concat(scalarmult(skA, otherPubKey), pkA, otherPubKey) or Hash(Concat(scalarmult(skA, otherPubKey), otherPubKey, pkA) depending on the order.

Reply:
> Since we are including an ECDH function to produce a shared secret we should point out that the shared secret as produced by scalarmult(., .) should NOT be used directly for encryption. Instead the output of ECDH should be concatenated with the bytes of the public keys and then hashed to produce a new secret.
> 
> > secret = Hash(concatenation(ECDH(skA, pkB), pkA, pkB))
> 
> (Where skA and pkA are of course produced following the BIP39/ARC-52 standard.)
> 
> Note that Alice and Bob need to agree beforehand on whose public key is concatenated first and whose second.
> 
> An explanation for why is given here: [jedisct1/libsodium#586 (comment)](https://github.com/jedisct1/libsodium/issues/586#issuecomment-325182090)
> 
> Scalarmult produces a point, a group element. While we use 32 bytes, corresponding to 2^256 possible values, in truth the order of the group is roughly ~2^252 and the output is not "uniformly random".
> 
> Hashing (e.g. with SHA512/256) makes the output uniformly random, but while the output would occupy the full 32 bytes space the range of inputs to the hash function is still constrained by ~2^252 possible inputs and the same number of outputs since hashes map (input) data to (output) value. By concatenating with something extra increase the range of outputs.
> 
> Keep in mind that the different combinations of n,p in scalarmult produce the same new point as well. Since Alice's pk is aG and Bob's bG, their shared secret is abG. a = 10, b = 50 would produce the same shared secret as a=20 and b = 25. However, Hash(Concat(150G, 10G, 50G)) !== Hash(Concat(150G, 20G, 25G)).
> 
> **Alternatively** we should update the ECDH so it doesn't return scalarmult(skA, pkB) directly. Instead we update it with an additional parameter "input order" and then returns either Hash(Concat(scalarmult(skA, otherPubKey), pkA, otherPubKey)) or Hash(Concat(scalarmult(skA, otherPubKey), otherPubKey, pkA)) depending on the order.

Great find. Update is needed then

Reply:
Also mentioned here: https://github.com/jedisct1/libsodium/blob/master/src/libsodium/include/sodium/crypto_scalarmult.h#L33

In crypto_kx() https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_kx/crypto_kx.c#L17 they're calling on "crypto_generichash".

Navigating to it shows that they're using blake2b: https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_generichash/crypto_generichash.c#L71

Regarding the input order, they're specifically concatenating the client pk first over the server pk. It might be clear to the two entities engaging in which is which, and if not they can negotiate (perhaps by comparing the bits of the PK and sorting the PK before the upper PK).

Reply:
Updated